python/lighthouse/utils/execution.py

```diff
@@ -0,0 +1,239 @@
+"""
+Execution engine utility functions.
+"""
+import numpy as np
+import ctypes
+import os
+from mlir import ir
+from mlir.dialects.transform import interpreter as transform_interpreter
+from mlir.dialects import func, arith, scf, memref
+from mlir.execution_engine import ExecutionEngine
+from mlir.runtime.np_to_memref import get_ranked_memref_descriptor
+from lighthouse.utils.mlir import get_mlir_library_path
+from lighthouse.utils import get_packed_arg
+from lighthouse import Workload
+from typing import Optional
+
+
+def get_engine(payload_module, requirements=None, opt_level=3) -> ExecutionEngine:
+    requirements = requirements or []
+    context = ir.Context()
+    location = ir.Location.unknown(context)
+    required_libs = {
+        "levelzero": (
+            ["libmlir_levelzero_runtime.so"],
+            "Did you compile LLVM with -DMLIR_ENABLE_LEVELZERO_RUNNER=1?",
+        ),
+        "mlir_runner": (["libmlir_runner_utils.so"], ""),
+        "mlir_c_runner": (["libmlir_c_runner_utils.so"], ""),
+    }
+    libs = []
+    lib_dir = os.path.join(get_mlir_library_path())
+    for r in requirements:
+        if r not in required_libs:
+            raise ValueError(f"Unknown execution engine requirement: {r}")
+        so_files, hint = required_libs[r]
+        for f in so_files:
+            so_path = os.path.join(lib_dir, f)
+            if not os.path.isfile(so_path):
+                msg = f"Could not find shared library {so_path}"
+                if hint:
+                    msg += "\n" + hint
+                raise ValueError(msg)
+            libs.append(so_path)
+    with context, location:
+        execution_engine = ExecutionEngine(
+            payload_module, opt_level=opt_level, shared_libs=libs
+        )
+        execution_engine.initialize()
+    return execution_engine
+
+
+def apply_transform_schedule(
+    payload_module,
+    schedule_module,
+    context,
+    location,
+    dump_kernel: Optional[str] = None,
+    dump_schedule: bool = False,
+):
+    if not dump_kernel or dump_kernel != "initial":
+        with context, location:
+            # invoke transform interpreter directly
+            transform_interpreter.apply_named_sequence(
```

Comment:
Now that we have https://github.com/llvm/llvm-project/pull/168223, this can probably be simplified. (E.g. do we need `apply_transform_schedule`?)